### PR TITLE
[PS-1824] Minor performance improvements in ObfuscateEmail

### PR DESCRIPTION
**Pull request by stevenaw**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

A few minor performance improvements in:
- `ApiHelpers.HandleAzureEvents()`
- `CoreHelpers.ObfuscateEmail()`

## Code changes

* **ApiHelpers.HandleAzureEvents():** Removed a duplicate dictionary lookup by converting a `ContainsKey()` + indexer call to instead use `TryGetValue()`
* **CoreHelpers.ObfuscateEmail():** Convert the obfuscation work from using `StringBuilder` to be span-based and use `string.Create()`

```BenchmarkDotNet=v0.13.2, OS=Windows 11 (10.0.22000.1098/21H2)
Intel Core i7-8565U CPU 1.80GHz (Whiskey Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK=7.0.100-rc.2.22477.23
  [Host]     : .NET 6.0.10 (6.0.1022.47605), X64 RyuJIT AVX2
  DefaultJob : .NET 6.0.10 (6.0.1022.47605), X64 RyuJIT AVX2


|             Method |                Email |      Mean |    Error |   StdDev |    Median |   Gen0 | Allocated |
|------------------- |--------------------- |----------:|---------:|---------:|----------:|-------:|----------:|
|      Obfuscate_New |       name@email.com |  82.22 ns | 1.721 ns | 4.768 ns |  83.34 ns | 0.0401 |     168 B |
| Obfuscate_Original |       name@email.com | 121.14 ns | 2.498 ns | 6.033 ns | 125.14 ns | 0.0725 |     304 B |
|      Obfuscate_New | reall(...)email [39] |  89.22 ns | 1.836 ns | 3.666 ns |  90.62 ns | 0.0631 |     264 B |
| Obfuscate_Original | reall(...)email [39] | 262.00 ns | 4.246 ns | 3.315 ns | 262.34 ns | 0.1392 |     584 B |
```

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: src/Core/Utilities/CoreHelpers.cs [hunk 1]
```diff
@@ -798,16 +798,21 @@ public static string ObfuscateEmail(string email)
             return email;
         }
 
-        var sb = new StringBuilder();
```
Comment (stevenaw) [inline src/Core/Utilities/CoreHelpers.cs:801]: NOTE: I didn't add unit tests for this change as some already existed, but please let me know if you'd prefer a few more

**Comment by justindbaur** [comment]
@stevenaw Thank you so much for this. I love this stuff! If you're up for it I think we can get rid of even more allocations. We could switch the `email.Split('@', ...)` to use a little helper I added here: https://github.com/bitwarden/server/blob/363dd6493a3996e1b9ef8522504259e60f4d7f31/src/Core/Utilities/SpanExtensions.cs#L5-L20

If you're up for that let me know but if not you can just sign the CLA and I can work on getting this merged.

File: src/Core/Utilities/CoreHelpers.cs [hunk 2]
```diff
@@ -798,16 +798,21 @@ public static string ObfuscateEmail(string email)
...
-        sb.Append(emailParts[0][..2]);
-        for (var i = 2; i < emailParts[0].Length; i++)
+        var finalLength = emailParts[0].Length + 1 + emailParts[1].Length;
```
Comment (justindbaur) [inline src/Core/Utilities/CoreHelpers.cs:801]: If you're up for it I think we can get rid of even more allocations. We could switch the `email.Split('@', ...)` to use a little helper I added here: https://github.com/bitwarden/server/blob/363dd6493a3996e1b9ef8522504259e60f4d7f31/src/Core/Utilities/SpanExtensions.cs#L5-L20

If you're up for that let me know but if not you can just sign the CLA and I can work on getting this merged.

**Review by justindbaur — COMMENTED** [review]
Thank you so much, I love these kind of improvements.

File: src/Core/Utilities/CoreHelpers.cs (hunk 2 above)
Comment (stevenaw) [reply src/Core/Utilities/CoreHelpers.cs:801]: Sure! Let me take another attempt. I wasn't sure whether the `'Split(char, StringSplitOptions.RemoveEmptyEntries)` was there to handle any edge cases but thinking through it a bit more I should be able to make some more improvements.

**Comment by stevenaw** [comment]
Ok, second attempt seems to have gone well. I was able to avoid the split altogether and simplify the code a bit by just writing the initial string into the new string's buffer and then overwriting the relevant part with a `Fill()` to make it effectively only two write operations.

```
|             Method |               Email |      Mean |     Error |     StdDev |    Median | Ratio | RatioSD |   Gen0 | Allocated | Alloc Ratio |
|------------------- |-------------------- |----------:|----------:|-----------:|----------:|------:|--------:|-------:|----------:|------------:|
| Obfuscate_Original |      name@email.com |  96.82 ns |  1.950 ns |   4.893 ns |  98.90 ns |  1.00 |    0.00 | 0.0726 |     304 B |        1.00 |
|       Obfuscate_V1 |      name@email.com |  73.33 ns |  1.530 ns |   3.753 ns |  74.73 ns |  0.76 |    0.05 | 0.0401 |     168 B |        0.55 |
|       Obfuscate_V2 |      name@email.com |  27.92 ns |  0.601 ns |   1.475 ns |  28.10 ns |  0.29 |    0.02 | 0.0344 |     144 B |        0.47 |
|                    |                     |           |           |            |           |       |         |        |           |             |
| Obfuscate_Original | real(...)mail [105] | 468.61 ns | 47.720 ns | 140.705 ns | 413.17 ns |  1.00 |    0.00 | 0.2637 |    1104 B |        1.00 |
|       Obfuscate_V1 | real(...)mail [105] | 195.98 ns |  4.218 ns |  12.103 ns | 196.15 ns |  0.46 |    0.12 | 0.1261 |     528 B |        0.48 |
|       Obfuscate_V2 | real(...)mail [105] |  77.34 ns |  1.654 ns |   2.982 ns |  76.73 ns |  0.21 |    0.02 | 0.0764 |     320 B |        0.29 |
```

This should be ready for another look @justindbaur, thanks for your prompt review last time. Happy to have the chance to contribute.

**Comment by stevenaw** [comment]
This seems to have stalled out. Anything I can do to help it along?

**Pull request closed without merging**